grafana/grafana
# Dashboard Scene:  Fix snapshots not displaying variables values

**What is this PR doing ?**
It fixes an issue for snapshots with template variables.  The variables were either not displayed at all or wrongly displayed.

**The issue** In snapshots, we only store the current variable value, and don't store `options` for variables such as `queries, datasources, ...` . The problem was that the Scenes implementation does validate `options` against the `current` variable value, if the `current` value did not exist, it cleared up that value. 

I created a new `SnapshotVariable`  and made some adjustments in the scenes repository ([PR](https://github.com/grafana/scenes/pull/781)) to support read-only for variables.
However, for `adhoc` variable types, I didn't use the snapshotVariable, as it was complex to add the same behavior to the custom snapshot variable, and the filters were not available.

![SnapshotVariables](https://github.com/grafana/grafana/assets/239999/23ad9f45-5aa2-4dcf-b146-00a72317abaa)




**How to test this PR?**
Create a snapshot with variables from different types and double-check check everything is working.

**Which issue(s) does this PR fix?**:



Fixes #85654

- [x] Add new `snapshot` type to Variable Type 
- [x] Remove console.logs
- [x] Add unit test for extra case 
- [x]  Extract variable logic outside of transformSaveModelToScene it was getting very confusing
